contracts/non-fungible-tokens/andromeda-crowdfund/src/contract.rs

```diff
@@ -288,6 +300,151 @@ fn execute_start_campaign(
+    match from_json(&cw20_msg.msg)? {
+        Cw20HookMsg::PurchaseTiers { orders } => {
+            purchase_tiers(ctx, Asset::Cw20Token(denom), amount, sender, orders)
+        }
+    }
+}
+
+fn purchase_tiers(
+    ctx: ExecuteContext,
+    denom: Asset,
+    amount: Uint128,
+    sender: String,
+    orders: Vec<SimpleTierOrder>,
+) -> Result<Response, ContractError> {
+    let ExecuteContext { deps, env, .. } = ctx;
+
+    let campaign_config = get_config(deps.storage)?;
+    let mut current_cap = get_current_cap(deps.storage);
+
+    let current_stage = get_current_stage(deps.storage);
+
+    // Tier can be purchased on ONGOING stage
+    ensure!(
+        current_stage == CampaignStage::ONGOING,
+        ContractError::InvalidCampaignOperation {
+            operation: "purchase_tiers".to_string(),
+            stage: current_stage.to_string()
+        }
+    );
+
+    // Need to wait until start_time
+    ensure!(
+        campaign_config
+            .start_time
+            .unwrap_or_default()
+            .is_expired(&env.block),
+        ContractError::CampaignNotStarted {}
+    );
+
+    // Campaign is expired or should be ended due to overfunding
+    ensure!(
+        !campaign_config.end_time.is_expired(&env.block)
+            || campaign_config.hard_cap.unwrap_or(current_cap) > current_cap,
+        ContractError::CampaignEnded {}
+    );
+
+    // Ensure campaign accepting coin is received
+    ensure!(
+        denom == campaign_config.denom,
+        ContractError::InvalidFunds {
+            msg: format!(
+                "Only {} is accepted by the campaign.",
+                campaign_config.denom
+            ),
+        }
+    );
+
+    let mut full_orders = Vec::<TierOrder>::new();
+
+    // Measure the total cost for orders
+    let total_cost = orders.iter().try_fold(Uint128::zero(), |sum, order| {
+        let tier = get_tier(deps.storage, u64::from(order.level))?;
+        let uint128: Result<Uint128, ContractError> = Ok(sum + tier.price * order.amount);
+        full_orders.push(TierOrder {
+            orderer: Addr::unchecked(sender.clone()),
+            level: order.level,
+            amount: order.amount,
+        });
+        uint128
+    })?;
+
+    // Ensure that enough payment is sent for the order
+    ensure!(total_cost <= amount, ContractError::InsufficientFunds {});
+
+    // Update order history and tier limits
+    set_tier_orders(deps.storage, full_orders)?;
+    current_cap = current_cap.checked_add(total_cost)?;
+
+    // Update current capital
+    set_current_cap(deps.storage, current_cap)?;
+    let mut resp = Response::new()
+        .add_attribute("action", "purchase_tiers")
+        .add_attribute("payment", format!("{0}{1}", amount, denom))
+        .add_attribute("total_cost", total_cost.to_string());
+
+    if amount > total_cost {
+        resp = match denom {
+            Asset::NativeToken(denom) => resp
+                .add_attribute("refunded", amount - total_cost)
```

Comment:
The attribute could be added to the response outside of the `match` statement since it's identical for both cases.

Reply:
Resolved